File: frontend/src/types/customer.ts
import { Address } from "./geographical";

export enum CustomerType {
  MARKETPLACE = 'MARKETPLACE',
  MERCHANT = 'MERCHANT'
}

export interface Customer {
  id: number;
  name: string;
  phone: string;
  alternativePhone?: string;
  facebookId?: string;
  address: Address;
  customerType: CustomerType;
  createdAt: string;
  updatedAt: string;
}

export interface CustomerRequest {
  id?: number | undefined;
  name: string;
  phone: string;
  divisionId: number;
  districtId: number;
  upazilaId: number;
  addressLine: string;
  postalCode?: string;
  alternativePhone?: string;
  facebookId?: string;
  customerType?: CustomerType;
}

File: src/main/java/com/tripzin/eleganttex/entity/Customer.java
package com.tripzin.eleganttex.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "customers")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Customer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "phone", nullable = false, unique = true)
    private String phone;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "address_id")
    private Address address;

    @Column(name = "alternative_phone")
    private String alternativePhone;

    @Column(name = "facebook_id")
    private String facebookId;

    @Enumerated(EnumType.STRING)
    @Column(name = "customer_type", nullable = false)
    private CustomerType customerType;

    @OneToMany(mappedBy = "customer")
    @Builder.Default
    private List<Order> orders = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    // Helper method to get formatted address
    public String getDisplayAddress() {
        if (address != null) {
            return address.getFormattedAddress();
        }
        return null;
    }
}
